Lessons page submenu items

LavaGolem opened the pull request:
Relates #4530

### Changes proposed in this Pull Request
* Add a tab for All Lessons that displays the default view.
* Add a tab for Lesson Tags that goes to the Lesson Tags page.
* Add an Order lessons link that goes to the Order Lessons page.
* Add a Lesson settings link that goes to the Lessons tab of Sensei LMS Settings.

### Testing instructions
* Go to the Lessons menu item in the Sensei menu 
* You should see this outlook of the menu
![image](https://user-images.githubusercontent.com/7208249/151337005-0b045d4c-6c13-4326-9187-ff20624b09da.png)
* **The first two items are tabs** and when clicking them you should see the following two views

| All lessons | Lesson tags | Lesson tag menu item |
| -- | -- | -- |
| ![image](https://user-images.githubusercontent.com/7208249/151337128-5d65330d-1547-47b2-adcc-c8858ef205f9.png) | ![image](https://user-images.githubusercontent.com/7208249/151337187-ac7290c1-e747-4df4-aeb0-3ee47038a8d9.png) | ![image](https://user-images.githubusercontent.com/7208249/151337298-fdfcb0d3-273f-48b7-a74a-cf393280f512.png) |


* Clicking New Lesson should redirect you to Lesson Creation view
![image](https://user-images.githubusercontent.com/7208249/151120365-101144eb-f092-47e0-8a93-dde89c26cb66.png)

* Clicking Order Lessons should lead you to 
![image](https://user-images.githubusercontent.com/7208249/151337436-87274e55-f5fc-48c3-9987-fa6447b0b509.png)
* Clicking Lesson Settings should lead you to 
![image](https://user-images.githubusercontent.com/7208249/151337487-15311383-9272-4317-b4d5-3149f2b0e021.png)

### New/Updated Hooks
Adding new Submenu item for Order Lessons so that that can be displayed in the Order Lessons tab
* add_action( 'admin_menu', array( $this, 'add_lesson_order' ) );

Adding custom navigation to the Lessons page 
* add_action( 'in_admin_header', [ $this, 'add_custom_navigation' ] );

Highlighting menu items 
*add_filter( 'submenu_file', [ $this, 'highlight_menu_item' ] );

merkushin commented on includes/class-sensei-admin.php:
```diff
@@ -348,9 +366,7 @@ public function register_scripts( $hook ) {
-		if ( $screen && ( in_array( $screen->id, [ 'edit-course', 'edit-course-category' ], true ) ) ) {
-			Sensei()->assets->enqueue( 'sensei-admin-custom-navigation', 'js/admin/custom-navigation.js', [], true );
-		}
+		Sensei()->assets->enqueue( 'sensei-admin-custom-navigation', 'js/admin/custom-navigation.js', [], true );
```
Previously @donnapep asked to enqueue these scripts only when they are needed: https://github.com/Automattic/sensei/pull/4672#discussion_r791904559

In my new PR with Modules navigation I extended this check this way: https://github.com/Automattic/sensei/pull/4693/files#diff-a700db88caca1edaa7ce24fc214d9344458fb9a1a636e4f8fa6bfb9eaa9a93dfR372-R375

LavaGolem replied:
I will then return the old if - logic and add it to the array, and when you merge you branch I can move it to your list, not to cause conflicts

merkushin reviewed (commented):
@LavaGolem looks great :)

I found two issues during testing.

1. We lose focus on the menu/submenu item when we go to "Lesson Tags", the main menu got collapsed.
![CleanShot 2022-01-27 at 16 30 21@2x](https://user-images.githubusercontent.com/329356/151370259-333980b9-0528-4018-9d78-2d53225a8c48.png)

2. I get an error after choosing a course in Order Lessons.

https://user-images.githubusercontent.com/329356/151370490-75d393b0-50ef-41dc-a988-34715355a198.mp4

merkushin commented on includes/class-sensei-lesson.php:
```diff
@@ -142,10 +142,77 @@ public function __construct() {
+	 * @since 4.0.0
+	 * @access private
+	 */
+	public function highlight_menu_item( ?string $submenu_file ) {
```
Sorry, didn't notice that during the review.
[Nullable types were introduced with PHP 7.1](https://www.php.net/manual/en/migration71.new-features.php)

donnapep replied:
This does indeed break everything on PHP 7.0 😅 :
![Screen Shot 2022-01-27 at 12 53 16 PM](https://user-images.githubusercontent.com/1190420/151415821-57cac37f-7c22-4ccd-a7c2-bbaa7cac6a78.jpg)

LavaGolem replied:
Fixed: https://github.com/Automattic/sensei/pull/4687/commits/b6e5dcbc1a59a9b20f955b6c360d0ec968e56377#diff-d8fb37cfbd6e262938ad7ebef55bb73e27759fb0b6e40d412985f5c7976243c5R174-R179

donnapep commented on includes/class-sensei-admin.php:
```diff
@@ -130,6 +131,22 @@ public function add_course_order() {
+			null,
+			__( 'Order Lessons', 'sensei-lms' ),
+			__( 'Order Lessons', 'sensei-lms' ),
+			'manage_sensei',
```
This should probably be `edit_published_lessons` [as it is currently](https://github.com/Automattic/sensei/blob/4aa6c35cce896ebb38f2231a3bbfc2ed40857a80/includes/class-sensei-admin.php#L112).

LavaGolem replied:
Fix: https://github.com/Automattic/sensei/pull/4687/commits/b6e5dcbc1a59a9b20f955b6c360d0ec968e56377#diff-a700db88caca1edaa7ce24fc214d9344458fb9a1a636e4f8fa6bfb9eaa9a93dfR146

donnapep commented on includes/class-sensei-admin.php:
```diff
@@ -130,6 +131,22 @@ public function add_course_order() {
+	/**
+	 * Add Lesson order page to admin panel.
+	 *
+	 * @since  4.0.0
```
We should mark this `@access private` so that other devs don't try to use it (since we're only making it public out of necessity for the hook to work).

LavaGolem replied:
fix: https://github.com/Automattic/sensei/pull/4687/commits/b6e5dcbc1a59a9b20f955b6c360d0ec968e56377#diff-d8fb37cfbd6e262938ad7ebef55bb73e27759fb0b6e40d412985f5c7976243c5R174-R179

donnapep commented on includes/class-sensei-lesson.php:
```diff
@@ -142,10 +142,77 @@ public function __construct() {
+	 *
+	 * @param string | null $submenu_file The submenu file points to the certain item of the submenu.
+	 *
+	 * @return string | null
```
Small thing, but the `@param`s and `@return` are usually the last thing, so below `@since` and `@access`.

LavaGolem replied:
fix: https://github.com/Automattic/sensei/pull/4687/commits/b6e5dcbc1a59a9b20f955b6c360d0ec968e56377#diff-d8fb37cfbd6e262938ad7ebef55bb73e27759fb0b6e40d412985f5c7976243c5R174-R179

donnapep commented on includes/class-sensei-lesson.php:
```diff
@@ -142,10 +142,77 @@ public function __construct() {
+					<h1><?php esc_html_e( 'Lessons', 'sensei-lms' ); ?></h1>
+				</div>
+				<div class="sensei-custom-navigation__links">
+					<a class="page-title-action" href="<?php echo esc_url( admin_url( 'post-new.php?post_type=lesson' ) ); ?>"><?php esc_html_e( 'New lesson', 'sensei-lms' ); ?></a>
```
As per the [Slack discussion the other day](p1643171662142100-slack-C02NWDZBL0H), we decided to use initial caps for "New Lesson", "Order Lessons" and "Lesson Settings".

LavaGolem replied:
@donnapep oh, taught this was only concerning the Tabs and not the links, since in both designs the links were with non-capitalized. Sure will change it

LavaGolem replied:
Fix: https://github.com/Automattic/sensei/pull/4687/commits/2e7de6bf89de705fe736bd76ec57abc68eb08593

donnapep commented:
> We lose focus on the main menu item when we go to "Lesson Tags", and the main menu got collapsed.

@merkushin Did you do anything special to solve this problem for "Course Categories"?

donnapep reviewed (commented):
Looking good! I left a few comments.

LavaGolem commented:
> @LavaGolem looks great :)
> 
> I found two issues during testing.
> 
> 1. We lose focus on the main menu item when we go to "Lesson Tags", and the main menu got collapsed.
>    ![CleanShot 2022-01-27 at 16 30 21@2x](https://user-images.githubusercontent.com/329356/151370259-333980b9-0528-4018-9d78-2d53225a8c48.png)
>    Not sure if it is a correct behavior.
> 2. I get an error after choosing a course in Order Lessons.
> 
>  CleanShot.2022-01-27.at.16.41.07.mp4

@merkushin  so regarding the second issue, it looks like this issue was present before I've found why and I've fixed it int this commit 

> @LavaGolem looks great :)
> 
> I found two issues during testing.
> 
> 1. We lose focus on the main menu item when we go to "Lesson Tags", and the main menu got collapsed.
>    ![CleanShot 2022-01-27 at 16 30 21@2x](https://user-images.githubusercontent.com/329356/151370259-333980b9-0528-4018-9d78-2d53225a8c48.png)
>    Not sure if it is a correct behavior.
> 2. I get an error after choosing a course in Order Lessons.
> 
>  CleanShot.2022-01-27.at.16.41.07.mp4

So on the second issue. Looks like the issue was present here from before. I've fixed it in this comment. https://github.com/Automattic/sensei/pull/4687/commits/72ef2196b551f34b1614d4a38e37d216ce5aa39d I'm not sure how to write a test for this but I guess this should be covered with integration tests. @donnapep  do you know what should we do with it?

donnapep commented on includes/class-sensei-lesson.php:
```diff
@@ -142,10 +142,78 @@ public function __construct() {
+			</div>
+			<div class="sensei-custom-navigation__tabbar">
+				<a class="sensei-custom-navigation__tab <?php echo '' === $screen->taxonomy ? 'active' : ''; ?>" href="<?php echo esc_url( admin_url( 'edit.php?post_type=lesson' ) ); ?>"><?php esc_html_e( 'All Lessons', 'sensei-lms' ); ?></a>
+				<a class="sensei-custom-navigation__tab <?php echo 'lesson-tag' === $screen->taxonomy ? 'active' : ''; ?>" href="<?php echo esc_url( admin_url( 'edit-tags.php?taxonomy=lesson-tag&post_type=lesson' ) ); ?>"><?php esc_html_e( 'Lesson Tags', 'sensei-lms' ); ?></a>
```
If we change `echo esc_url( admin_url( 'edit-tags.php?taxonomy=lesson-tag&post_type=lesson' )` to `echo esc_url( admin_url( 'edit-tags.php?taxonomy=lesson-tag&post_type=course' ), it should keep the _Lessons_ menu highlighted when viewing _Lesson Tags_.

donnapep reviewed (approved):
Works well. 🎉  I suggested a fix for keeping the correct menu highlighted when viewing the lesson tags.

LavaGolem commented:
> Works well. 🎉 I suggested a fix for keeping the correct menu highlighted when viewing the lesson tags.

I just pushed one commit, with linter fix: https://github.com/Automattic/sensei/pull/4687/commits/3131100931f7acbed9addf6649a44151e69aa539
